test/lib/task_manager.spec.js

```diff
@@ -0,0 +1,179 @@
+const { assert, expect } = require('chai');
+const TaskManager = require('../../lib/task_manager');
+const { MAX_TASK_DATA_KEYS, START_BLOCK, START_PRIMARY_KEY } = require('../../lib/constants');
+
+describe('TaskManager', () => {
+  it('constructor initializes corresponding variables', () => {
+    const taskManager = new TaskManager();
+
+    assert.deepStrictEqual(taskManager.taskData, {});
+    assert.strictEqual(taskManager.consequentTaskIndex, 0);
+
+    assert.strictEqual(taskManager.queue, undefined);
+    assert.strictEqual(taskManager.lastPushedToBuffer, undefined);
+    assert.strictEqual(taskManager.lastPrimaryKey, undefined);
+    assert.strictEqual(taskManager.lastExportedBlock, undefined);
+  });
+
+  it('queue is initialized accordingly', async () => {
+    const taskManager = new TaskManager();
+    await taskManager.initQueue(5);
+
+    assert.isDefined(taskManager.queue);
+    assert.strictEqual(taskManager.queue.concurrency, 5);
+  });
+
+  it('create static method initializes the TaskManager with its queue', async () => {
+    const CONCURRENCY = 2;
+    const taskManager = await TaskManager.create(CONCURRENCY);
+    const PQueue = (await import('p-queue')).default;
+
+    expect(taskManager.queue).to.be.an.instanceof(PQueue);
+    expect(taskManager.queue.concurrency).to.eq(CONCURRENCY);
+
+    expect(taskManager).to.be.an.instanceof(TaskManager);
+  });
+
+  it('queue handles tasks accordingly when they finish', async () => {
+    const CONCURRENCY = 1;
+    const taskManager = await TaskManager.create(CONCURRENCY);
+
+    const exampleEventsData = [ 1, 2, 3, 4, 5 ];
+    const emmitedTaskData = [{ fromBlock: 1, toBlock: 10 }, exampleEventsData];
+    taskManager.queue.emit('completed', emmitedTaskData);
+
+    expect(taskManager.taskData).to.deep.eq({ 1: { toBlock: 10, data: exampleEventsData } });
+    expect(taskManager.consequentTaskIndex).to.eq(1);
+    expect(taskManager.queue.isPaused).to.eq(false);
+  });
+
+  it('queue is paused when `consequentTaskIndex` crosses the borderline', async () => {
+    const CONCURRENCY = 1;
+    const taskManager = await TaskManager.create(CONCURRENCY);
+    taskManager.consequentTaskIndex = MAX_TASK_DATA_KEYS;
+
+    const exampleEventsData = [ 1, 2, 3, 4, 5 ];
+    const emmitedTaskData = [{ fromBlock: 1, toBlock: 10 }, exampleEventsData];
+    taskManager.queue.emit('completed', emmitedTaskData);
+
+    expect(taskManager.queue.isPaused).to.eq(true);
```

Comment:
I don't think that test reflects the implementation, I think all the TaskManager tests need revisiting. In the current implementation the TaskManager is a nice standalone implementation which can be black box tested.